File: flutter/lib/screens/currency_converter.dart
// Currency converter: type an amount, pick From and To, see what it is worth.
// A reference tool only, it shows values and never moves, trades, or sends
// money. Rates come from the internet when online and are cached for offline;
// only the base currency code is ever sent. All the money math is the
// golden-locked crossRate/roundRate/formatConverted from the pure core.

import 'package:flutter/material.dart';
import 'package:flutter/services.dart' show FilteringTextInputFormatter;

import '../data/fx_service.dart';
import '../data/store.dart';
import '../money/currencies.dart';
import '../money/fxrates.dart';
import '../theme.dart';
import '../typography.dart';
import '../widgets/salapify_icon.dart';

const List<String> _mos = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
];

String _asOf(int? ms) {
  if (ms == null || ms <= 0) return '';
  final d = DateTime.fromMillisecondsSinceEpoch(ms);
  return '${_mos[d.month - 1]} ${d.day}, ${d.year}';
}

class CurrencyConverterScreen extends StatefulWidget {
  final SalapifyStore store;
  const CurrencyConverterScreen({super.key, required this.store});

  @override
  State<CurrencyConverterScreen> createState() =>
      _CurrencyConverterScreenState();
}

class _CurrencyConverterScreenState extends State<CurrencyConverterScreen> {
  late final String _base;
  final _amount = TextEditingController();
  late String _from;
  late String _to;
  FxRates? _fx;
  bool _loading = true;

  @override
  void initState() {
    super.initState();
    _base = _baseCode(widget.store.data['settings']);
    _from = _base;
    _to = _base == 'USD' ? 'PHP' : 'USD';
    _amount.addListener(() => setState(() {}));
    _load();
  }

  String _baseCode(dynamic settings) {
    if (settings is Map) {
      for (final key in ['currencyCode', 'currency']) {
        final v = settings[key];
        if (v is String && currencies.any((c) => c['code'] == v)) return v;
      }
    }
    return 'PHP';
  }

  Future<void> _load() async {
    final fx = await FxService().load(_base);
    if (!mounted) return;
    setState(() {
      _fx = fx;
      _loading = false;
    });
  }

  @override
  void dispose() {
    _amount.dispose();
    super.dispose();
  }

  void _swap() => setState(() {
    final t = _from;
    _from = _to;
    _to = t;
  });

  void _retry() {
    setState(() => _loading = true);
    _load();
  }

  @override
  Widget build(BuildContext context) {
    final amountNum =
        double.tryParse(_amount.text.replaceAll(RegExp(r'[, ]'), '')) ?? 0;
    final haveRates = _fx != null && _fx!.base == _base;
    final rate = haveRates ? crossRate(_fx!.rates, _from, _to) : null;
    final converted = convertAmount(amountNum, rate);
    final asOf = haveRates ? _asOf(_fx!.fetchedAt) : '';

    return Scaffold(
      appBar: AppBar(
        backgroundColor: Barako.background,
        foregroundColor: Barako.text,
        title: Text(
          'Currency converter',
          style: TextStyle(color: Barako.text, fontWeight: FontWeight.w800),
        ),
      ),
      body: SafeArea(
        child: ListView(
          padding: const EdgeInsets.fromLTRB(20, 8, 20, 32),
          children: [
            Text(
              'See what your money is worth in another currency. Rates come from the internet when you are online and are saved for offline use. This shows values only, it does not exchange or move money.',
              style: AppText.small.tint(Barako.muted).copyWith(height: 1.5),
            ),
            const SizedBox(height: 18),
            Text('AMOUNT IN $_from', style: Barako.kickerStyle),
            const SizedBox(height: 6),
            Row(
              children: [
                Text(
                  currencySymbol(_from),
                  style: AppText.heading.w4.tint(Barako.textSecondary),
                ),
                const SizedBox(width: 8),
                Expanded(
                  child: TextField(
                    controller: _amount,
                    keyboardType: const TextInputType.numberWithOptions(
                      decimal: true,
                    ),
                    inputFormatters: [
                      FilteringTextInputFormatter.allow(RegExp(r'[0-9., ]')),
                    ],
                    autofocus: true,
                    style: AppText.title.w7,
                    decoration: InputDecoration(
                      hintText: '0',
                      hintStyle: TextStyle(color: Barako.faint),
                    ),
                  ),
                ),
              ],
            ),
            const SizedBox(height: 16),
            _currencyRow('FROM', _from, (c) => setState(() => _from = c)),
            Center(
              child: TextButton.icon(
                onPressed: _swap,
                icon: Icon(
                  salapifyIcon('swap'),
                  color: Barako.primary,
                  size: 18,
                ),
                label: Text(
                  'Swap',
                  style: TextStyle(
                    color: Barako.primary,
                    fontWeight: FontWeight.w700,
                  ),
                ),
              ),
            ),
            _currencyRow('TO', _to, (c) => setState(() => _to = c)),
            const SizedBox(height: 18),
            _resultCard(amountNum, converted, rate, asOf),
            const SizedBox(height: 18),
            Center(
              child: Text(
                'Rates by Exchange Rate API',
                style: AppText.micro.w4.tint(Barako.faint),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _currencyRow(
    String label,
    String value,
    void Function(String) onPick,
  ) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(label, style: Barako.kickerStyle),
        const SizedBox(height: 8),
        SizedBox(
          // At least a 44dp tap target, and it grows with the system text
          // scale so the pills never clip for large-font users.
          height:
              48 * MediaQuery.textScalerOf(context).scale(1).clamp(1.0, 1.6),
          child: ListView.separated(
            scrollDirection: Axis.horizontal,
            itemCount: currencies.length,
            separatorBuilder: (_, _) => const SizedBox(width: 8),
            itemBuilder: (_, i) {
              final c = currencies[i];
              final code = c['code']!;
              final on = value == code;
              return ChoiceChip(
                label: Text('${c['symbol']} $code'),
                selected: on,
                onSelected: (_) => onPick(code),
                selectedColor: Barako.primary,
                backgroundColor: Barako.background,
                labelStyle: TextStyle(
                  color: on ? Barako.onPrimary : Barako.textSecondary,
                  fontWeight: FontWeight.w600,
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _resultCard(
    double amountNum,
    double? converted,
    double? rate,
    String asOf,
  ) {
    Widget body;
    if (_from == _to) {
      body = Column(
        children: [
          _big(formatConverted(amountNum, _to)),
          const SizedBox(height: 8),
          Text(
            'Same currency, nothing to convert.',
            textAlign: TextAlign.center,
            style: AppText.caption,
          ),
        ],
      );
    } else if (converted != null && rate != null) {
      body = Column(
        children: [
          _big(formatConverted(converted, _to)),
          const SizedBox(height: 8),
          Text(
            '1 $_from = ${roundRate(rate)} $_to${asOf.isNotEmpty ? ' · rates as of $asOf' : ''}',
            textAlign: TextAlign.center,
            style: AppText.caption,
          ),
        ],
      );
    } else if (_loading) {
      body = Text("Getting today's rates...", style: AppText.small);
    } else if (_fx != null) {
      // Rates are loaded, this pair just is not covered by the table. Do not
      // tell an online user to connect.
      body = Text(
        'No rate for $_from to $_to right now. Try another currency.',
        textAlign: TextAlign.center,
        style: AppText.small.copyWith(height: 1.4),
      );
    } else {
      body = Column(
        children: [
          Text(
            "No rates yet. Connect to the internet once to download today's rates, then it works offline too.",
            textAlign: TextAlign.center,
            style: AppText.small.copyWith(height: 1.4),
          ),
          const SizedBox(height: 8),
          TextButton(
            onPressed: _retry,
            child: Text(
              'Try again',
              style: TextStyle(
                color: Barako.primary,
                fontWeight: FontWeight.w700,
              ),
            ),
          ),
        ],
      );
    }
    return Card(
      child: Padding(
        padding: const EdgeInsets.all(24),
        child: Center(child: body),
      ),
    );
  }

  Widget _big(String text) => FittedBox(
    fit: BoxFit.scaleDown,
    child: Text(
      text,
      maxLines: 1,
      style: AppText.amountXl.tint(Barako.primary),
    ),
  );
}

File: mobile/app/currency-converter.js
// Currency converter. A simple reference tool: type an amount, pick From and To,
// see what it is worth. It reuses the same live rate layer as logging a foreign
// expense (hooks/useFxRates), so rates are fetched when online and cached for
// offline, and only your base currency code is ever sent, never your data.
//
// Deliberately a CONVERTER, not an exchange: it shows what money is worth, it
// never moves, trades, or sends money. No dashes anywhere per house style.

import { useMemo, useState } from 'react';
import { Pressable, ScrollView, StyleSheet, Text, TextInput, View, Linking } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { useRouter } from 'expo-router';
import { Ionicons } from '@expo/vector-icons';
import { spacing, radius, fontSize, fontWeight } from '../theme';
import { useTheme } from '../context/Theme';
import { useAppData } from '../context/AppData';
import { CURRENCIES, currencySymbol, formatConverted } from '../lib/currencies';
import { useFxRates } from '../hooks/useFxRates';
import { crossRate, roundRate } from '../lib/fxrates';

// Hoisted out of the screen so it keeps its identity across renders: defining it
// inline would remount both chip rows on every keystroke and snap them back to
// the start, hiding the currency the user picked.
function CurrencyRow({ label, value, onPick, styles }) {
  return (
    <>
      <Text style={styles.fieldLabel}>{label}</Text>
      <ScrollView horizontal showsHorizontalScrollIndicator={false} contentContainerStyle={styles.chips}>
        {CURRENCIES.map((c) => {
          const on = value === c.code;
          return (
            <Pressable key={c.code} onPress={() => onPick(c.code)} style={[styles.chip, on && styles.chipOn]}>
              <Text style={[styles.chipText, on && styles.chipTextOn]}>{c.symbol} {c.code}</Text>
            </Pressable>
          );
        })}
      </ScrollView>
    </>
  );
}

export default function CurrencyConverter() {
  const { colors } = useTheme();
  const styles = useMemo(() => makeStyles(colors), [colors]);
  const router = useRouter();
  const { data } = useAppData();

  const baseCode = (data.settings && data.settings.currencyCode) || 'PHP';
  const [amount, setAmount] = useState('');
  const [from, setFrom] = useState(baseCode);
  const [to, setTo] = useState(baseCode === 'USD' ? 'PHP' : 'USD');

  // One rates table (units per base currency); cross rates convert between any two.
  const fx = useFxRates(baseCode);
  const haveRates = fx.base === baseCode && fx.rates;
  const rate = haveRates ? crossRate(fx.rates, from, to) : null;

  const amountNum = Number(String(amount).replace(/[, ]/g, '')) || 0;
  const converted = rate != null ? amountNum * rate : null;

  // The rate table's date, so the user knows how fresh the numbers are.
  const asOf = (() => {
    if (!fx.fetchedAt) return '';
    try {
      const d = new Date(fx.fetchedAt);
      return d.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
    } catch (e) {
      return '';
    }
  })();

  const swap = () => {
    setFrom(to);
    setTo(from);
  };

  return (
    <SafeAreaView style={styles.screen} edges={['top']}>
      <View style={styles.headerBar}>
        <Pressable onPress={() => router.back()} hitSlop={10} style={styles.back} accessibilityRole="button" accessibilityLabel="Go back">
          <Ionicons name="chevron-back" size={24} color={colors.text} />
        </Pressable>
        <Text style={styles.headerTitle}>Currency converter</Text>
        <View style={{ width: 24 }} />
      </View>

      <ScrollView contentContainerStyle={styles.content} keyboardShouldPersistTaps="handled">
        <Text style={styles.intro}>
          See what your money is worth in another currency. Rates come from the internet when you are online and are saved for offline use. This shows values only, it does not exchange or move money.
        </Text>

        <Text style={styles.fieldLabel}>Amount in {from}</Text>
        <View style={styles.inputWrap}>
          <Text style={styles.sym}>{currencySymbol(from)}</Text>
          <TextInput
            style={styles.input}
            value={amount}
            onChangeText={setAmount}
            keyboardType="numeric"
            placeholder="0"
            placeholderTextColor={colors.faint}
            autoFocus
          />
        </View>

        <CurrencyRow label="From" value={from} onPick={setFrom} styles={styles} />

        <Pressable onPress={swap} style={styles.swapBtn} accessibilityRole="button" accessibilityLabel="Swap currencies">
          <Ionicons name="swap-vertical" size={18} color={colors.primary} />
          <Text style={styles.swapText}>Swap</Text>
        </Pressable>

        <CurrencyRow label="To" value={to} onPick={setTo} styles={styles} />

        <View style={styles.resultCard}>
          {from === to ? (
            <Text style={styles.resultBig}>{formatConverted(amountNum, to)}</Text>
          ) : converted != null ? (
            <>
              <Text style={styles.resultBig}>{formatConverted(converted, to)}</Text>
              <Text style={styles.resultRate}>
                1 {from} = {roundRate(rate)} {to}
                {asOf ? ` · rates as of ${asOf}` : ''}
              </Text>
            </>
          ) : fx.loading ? (
            <Text style={styles.resultNote}>Getting today's rates…</Text>
          ) : (
            <Text style={styles.resultNote}>
              No rate for {from} to {to} yet. Connect to the internet once to download today's rates, then it works offline too.
            </Text>
          )}
        </View>

        <Pressable onPress={() => Linking.openURL('https://www.exchangerate-api.com').catch(() => {})}>
          <Text style={styles.attribution}>Rates by Exchange Rate API</Text>
        </Pressable>
      </ScrollView>
    </SafeAreaView>
  );
}

function makeStyles(colors) {
  return StyleSheet.create({
    screen: { flex: 1, backgroundColor: colors.background },
    headerBar: { flexDirection: 'row', alignItems: 'center', justifyContent: 'space-between', paddingHorizontal: spacing.lg, paddingTop: spacing.md, paddingBottom: spacing.sm },
    back: { marginLeft: -4 },
    headerTitle: { color: colors.text, fontSize: fontSize.subtitle, fontWeight: fontWeight.bold },
    content: { padding: spacing.lg, paddingBottom: spacing.xxl },
    intro: { color: colors.muted, fontSize: fontSize.small, marginBottom: spacing.lg, lineHeight: 19 },

    fieldLabel: { color: colors.muted, fontSize: fontSize.caption, marginBottom: spacing.xs, marginTop: spacing.md },
    inputWrap: { flexDirection: 'row', alignItems: 'center', backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.md, paddingHorizontal: spacing.md },
    sym: { color: colors.textSecondary, fontSize: fontSize.body, marginRight: spacing.xs },
    input: { flex: 1, paddingVertical: spacing.md, color: colors.text, fontSize: fontSize.subtitle, fontWeight: fontWeight.bold },

    chips: { flexDirection: 'row', gap: spacing.sm, paddingVertical: 2 },
    chip: { paddingVertical: spacing.xs, paddingHorizontal: spacing.md, borderRadius: radius.pill, borderWidth: 1, borderColor: colors.border, backgroundColor: colors.card },
    chipOn: { backgroundColor: colors.primary, borderColor: colors.primary },
    chipText: { color: colors.text, fontSize: fontSize.small },
    chipTextOn: { color: colors.onPrimary, fontWeight: fontWeight.bold },

    swapBtn: { flexDirection: 'row', alignItems: 'center', gap: spacing.xs, alignSelf: 'center', marginTop: spacing.md, paddingVertical: spacing.xs, paddingHorizontal: spacing.md },
    swapText: { color: colors.primary, fontSize: fontSize.small, fontWeight: fontWeight.bold },

    resultCard: { backgroundColor: colors.card, borderColor: colors.border, borderWidth: 1, borderRadius: radius.lg, padding: spacing.xl, marginTop: spacing.lg, alignItems: 'center' },
    resultBig: { color: colors.primary, fontSize: fontSize.title, fontWeight: fontWeight.heavy },
    resultRate: { color: colors.muted, fontSize: fontSize.small, marginTop: spacing.sm, textAlign: 'center' },
    resultNote: { color: colors.textSecondary, fontSize: fontSize.small, textAlign: 'center', lineHeight: 18 },

    attribution: { color: colors.faint, fontSize: fontSize.caption, textAlign: 'center', marginTop: spacing.lg, textDecorationLine: 'underline' },
  });
}
